test/validation/api/ipsec/ipsec_test_in.c

```diff
@@ -1764,132 +1767,109 @@ static void test_ipsec_sa_print(void)
 	}
 }
 
-static void test_in_ipv4_esp_reass_success_four_frags(odp_ipsec_sa_t out_sa,
-						      odp_ipsec_sa_t in_sa)
+static void test_in_ipv4_esp_reass_success_two_frags(odp_ipsec_sa_t out_sa,
+						     odp_ipsec_sa_t in_sa)
 {
-	ipsec_test_part test_out[MAX_FRAGS], test_in[MAX_FRAGS];
-	int i;
-
-	memset(test_in, 0, sizeof(test_in));
-
-	CU_ASSERT(MAX_FRAGS >= 4);
-
-	part_prep_esp(test_out, 4, false);
-
-	test_out[0].pkt_in = &pkt_ipv4_udp_p2_f1;
-	test_out[1].pkt_in = &pkt_ipv4_udp_p2_f2;
-	test_out[2].pkt_in = &pkt_ipv4_udp_p2_f3;
-	test_out[3].pkt_in = &pkt_ipv4_udp_p2_f4;
-
-	part_prep_plain(&test_in[3], 1, false, true);
-	test_in[3].out[0].pkt_res = &pkt_ipv4_udp_p2;
-
-	for (i = 0; i < 4; i++) {
-		ipsec_test_packet test_pkt;
-		odp_packet_t pkt;
+	ipsec_test_packet *input_packets[] = {
+		&pkt_ipv4_udp_p1_f1,
+		&pkt_ipv4_udp_p1_f2,
+	};
+	ipsec_test_packet *result_packet = &pkt_ipv4_udp_p1;
 
-		CU_ASSERT_EQUAL(ipsec_check_out(&test_out[i], out_sa, &pkt), 1);
+	test_multi_out_in(out_sa, in_sa, ODPH_IPV4,
+			  ARRAY_SIZE(input_packets),
+			  input_packets,
+			  result_packet);
+}
 
-		ipsec_test_packet_from_pkt(&test_pkt, &pkt);
-		test_in[i].pkt_in = &test_pkt;
+static void test_in_ipv4_esp_reass_success_four_frags(odp_ipsec_sa_t out_sa,
+						      odp_ipsec_sa_t in_sa)
+{
+	ipsec_test_packet *input_packets[] = {
+		&pkt_ipv4_udp_p2_f1,
+		&pkt_ipv4_udp_p2_f2,
+		&pkt_ipv4_udp_p2_f3,
+		&pkt_ipv4_udp_p2_f4,
+	};
+	ipsec_test_packet *result_packet = &pkt_ipv4_udp_p2;
 
-		ipsec_check_in_one(&test_in[i], in_sa);
-	}
+	test_multi_out_in(out_sa, in_sa, ODPH_IPV4,
+			  ARRAY_SIZE(input_packets),
+			  input_packets,
+			  result_packet);
 }
 
 static void test_in_ipv4_esp_reass_success_two_frags_ooo(odp_ipsec_sa_t out_sa,
 							 odp_ipsec_sa_t in_sa)
 {
-	ipsec_test_part test_out[MAX_FRAGS], test_in[MAX_FRAGS];
-	int i;
-
-	memset(test_in, 0, sizeof(test_in));
-
-	CU_ASSERT(MAX_FRAGS >= 2);
-
-	part_prep_esp(test_out, 2, false);
-
-	test_out[0].pkt_in = &pkt_ipv4_udp_p1_f2;
-	test_out[1].pkt_in = &pkt_ipv4_udp_p1_f1;
-
-	part_prep_plain(&test_in[1], 1, false, true);
-	test_in[1].out[0].pkt_res = &pkt_ipv4_udp_p1;
-
-	for (i = 0; i < 2; i++) {
-		ipsec_test_packet test_pkt;
-		odp_packet_t pkt;
-
-		CU_ASSERT_EQUAL(ipsec_check_out(&test_out[i], out_sa, &pkt), 1);
-
-		ipsec_test_packet_from_pkt(&test_pkt, &pkt);
-		test_in[i].pkt_in = &test_pkt;
+	ipsec_test_packet *input_packets[] = {
+		&pkt_ipv4_udp_p1_f2,
+		&pkt_ipv4_udp_p1_f1,
+	};
+	ipsec_test_packet *result_packet = &pkt_ipv4_udp_p1;
 
-		ipsec_check_in_one(&test_in[i], in_sa);
-	}
+	test_multi_out_in(out_sa, in_sa, ODPH_IPV4,
+			  ARRAY_SIZE(input_packets),
+			  input_packets,
+			  result_packet);
 }
 
 static void test_in_ipv4_esp_reass_success_four_frags_ooo(odp_ipsec_sa_t out_sa,
 							  odp_ipsec_sa_t in_sa)
 {
-	ipsec_test_part test_out[MAX_FRAGS], test_in[MAX_FRAGS];
-	int i;
-
-	memset(test_in, 0, sizeof(test_in));
-
-	CU_ASSERT(MAX_FRAGS >= 4);
```

Comment:
@JannePeltonen I remember adding this ASSERT on your insistence. I'm not able to recall why that ASSERT was required. Can you check if that is required. We can put in test_multi_out_in() if required.

Reply:
The assert was there to check that the array defined in the function was large enough. IOW, the code in the function assumed that MAX_FRAGS was at least 4. The new common function does not assume that and therefore does not have the assert.

Now that I look at the code that was not changed, the MAX_FRAGS stuff and the assumptions around the pkto array sizes are a bit fragile and could use a few asserts here and there. But that is for some other patch. Maybe I will take a better look and submit a new PR about it.